### PR TITLE
Skeleton testing framework

### PR DESCRIPTION
**Pull request by texasmichelle**
First stab at adding tests to this repo.

@jlewi what's the best way to see if this runs before merging? How can I run it locally?

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/examples/18)

### CODE REVIEW

File: test/test-infra/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Dockerfile used by out prow jobs.
+# # The sole purpose of this image is to customize the command run.
+FROM gcr.io/mlkube-testing/test-worker:latest
```
Comment (jlewi) [inline test/test-infra/Dockerfile]: I don't think you need a Dockerfile. There's no reason we should be building our own imag.e

File: test/test-infra/Makefile [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2017 The Kubernetes Authors.
```
Comment (jlewi) [inline test/test-infra/Makefile]: I think you can delete everythin in test-infra. I don't think you need that directory.

**Comment by puneith** [comment]
@texasmichelle Have you looked into Kokoro for testing? We have it setup for Cloud ML Samples Github.

**Comment by jlewi** [comment]
Lets not use Jenkins or Kokoro.
Prow is K8s native; Jenkins is not.
Kubernetes made a concise decision to move away from Jenkins to Prow.

Kokoro is largely an internal tool to Google.

**Comment by texasmichelle** [comment]
The change count in this PR is high. Should we be ignoring test/workflows/environments/test/.metadata?

**Comment by jlewi** [comment]
For the most recent run the results gotten written to gubernator so you can click on details to get the gubernator page.

From there you can see that the workflow it ran for the E2E test failed and you can get an Argo link

http://testing-argo.kubeflow.io/timeline/kubeflow-test-infra/kubeflow-examples-presubmit-examples-e2e-18-57372ed-19-97dd;tab=workflow

The bulld step, lint step, and test step all failed.

Logs show

```
/usr/bin/python: No module named py
```

If you're goal in this PR is just to make the test "pass" then you could make the test a null op by just deleting those steps from the workflow.

Another option would be to copy the code from kubeflow/testing/py. But I'd prefer not to duplicate code.

My preference would be to update the scripts in kubeflow/testing/py  so that they can be used for other repositories (you'll need to do things like make the exclude directories command line arguments).

You can then use the environment variable in EXTRA_REPOS to check out kubeflow/testing. You could then use the scripts in kubeflow/testing to run unit and lint tests.

**Comment by jlewi** [comment]
It doesn't look like the workflow even ran.

http://testing-argo.kubeflow.io/timeline/kubeflow-test-infra/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a;tab=workflow

To debug this you would need access to the test cluster which I can grant.
At which point you could 

```
kubectl describe wf  kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a
```

This shows an error in the workflow spec

```
Name:         kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a
Namespace:    kubeflow-test-infra
Labels:       workflows.argoproj.io/completed=true
              workflows.argoproj.io/phase=Failed
Annotations:  <none>
API Version:  argoproj.io/v1alpha1
Kind:         Workflow
Metadata:
  Cluster Name:                   
  Creation Timestamp:             2018-02-28T00:13:42Z
  Deletion Grace Period Seconds:  <nil>
  Deletion Timestamp:             <nil>
  Generation:                     0
  Initializers:                   <nil>
  Resource Version:               7440636
  Self Link:                      /apis/argoproj.io/v1alpha1/namespaces/kubeflow-test-infra/workflows/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a
  UID:                            3c20c54d-1c1c-11e8-a0a6-42010af00080
Spec:
  Entrypoint:  e2e
  On Exit:     exit-handler
  Templates:
    Name:  e2e
    Steps:
      [map[name:checkout template:checkout]]
      [map[name:create-pr-symlink template:create-pr-symlink] map[name:py-test template:py-test] map[name:py-lint template:py-lint]]
    Container:
      Command:
        /usr/local/bin/checkout.sh
        /mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src
      Env:
        Name:   BUILD_ID
        Value:  21
        Name:   BUILD_NUMBER
        Value:  21
        Name:   JOB_NAME
        Value:  kubeflow-examples-presubmit
        Name:   JOB_TYPE
        Value:  presubmit
        Name:   PULL_BASE_SHA
        Value:  f4fad9a8e0f8d27dc2fa07b4a6cd792ff26634ed
        Name:   PULL_NUMBER
        Value:  18
        Name:   PULL_PULL_SHA
        Value:  68c34529a3b785d7f22804776c13f29160efdeaa
        Name:   REPO_NAME
        Value:  examples
        Name:   REPO_OWNER
        Value:  kubeflow
        Name:   EXTRA_REPOS
        Value:  kubeflow/testing@HEAD
      Image:    gcr.io/mlkube-testing/test-worker
      Volume Mounts:
        Mount Path:  /mnt/test-data-volume
        Name:        kubeflow-test-volume
    Name:            checkout
    Container:
      Command:
        python
        -m
        kubeflow.testing.py_checks
        test
        --src_dir=/mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/examples
        --project=mlkube-testing
        --junit_path=/mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/output/artifacts/junit_pycheckstest.xml
      Env:
        Name:   PYTHONPATH
        Value:  /mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/examples:/mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/testing/py
        Name:   GOPATH
        Value:  /mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/go
        Name:   GOOGLE_APPLICATION_CREDENTIALS
        Value:  /secret/gcp-credentials/key.json
        Name:   GIT_TOKEN
        Value From:
          Secret Key Ref:
            Key:   github_token
            Name:  github-token
        Name:      EXTRA_REPOS
        Value:     kubeflow/testing@HEAD
        Name:      BUILD_ID
        Value:     21
        Name:      BUILD_NUMBER
        Value:     21
        Name:      JOB_NAME
        Value:     kubeflow-examples-presubmit
        Name:      JOB_TYPE
        Value:     presubmit
        Name:      PULL_BASE_SHA
        Value:     f4fad9a8e0f8d27dc2fa07b4a6cd792ff26634ed
        Name:      PULL_NUMBER
        Value:     18
        Name:      PULL_PULL_SHA
        Value:     68c34529a3b785d7f22804776c13f29160efdeaa
        Name:      REPO_NAME
        Value:     examples
        Name:      REPO_OWNER
        Value:     kubeflow
      Image:       gcr.io/mlkube-testing/test-worker
      Volume Mounts:
        Mount Path:  /mnt/test-data-volume
        Name:        kubeflow-test-volume
        Mount Path:  /secret/github-token
        Name:        github-token
        Mount Path:  /secret/gcp-credentials
        Name:        gcp-credentials
      Working Dir:   /mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/examples
    Name:            py-test
    Container:
      Command:
        python
        -m
        kubeflow.testing.py_checks
        lint
        --src_dir=/mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/examples
        --project=mlkube-testing
        --junit_path=/mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/output/artifacts/junit_pycheckslint.xml
      Env:
        Name:   PYTHONPATH
        Value:  /mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/examples:/mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/testing/py
        Name:   GOPATH
        Value:  /mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/go
        Name:   GOOGLE_APPLICATION_CREDENTIALS
        Value:  /secret/gcp-credentials/key.json
        Name:   GIT_TOKEN
        Value From:
          Secret Key Ref:
            Key:   github_token
            Name:  github-token
        Name:      EXTRA_REPOS
        Value:     kubeflow/testing@HEAD
        Name:      BUILD_ID
        Value:     21
        Name:      BUILD_NUMBER
        Value:     21
        Name:      JOB_NAME
        Value:     kubeflow-examples-presubmit
        Name:      JOB_TYPE
        Value:     presubmit
        Name:      PULL_BASE_SHA
        Value:     f4fad9a8e0f8d27dc2fa07b4a6cd792ff26634ed
        Name:      PULL_NUMBER
        Value:     18
        Name:      PULL_PULL_SHA
        Value:     68c34529a3b785d7f22804776c13f29160efdeaa
        Name:      REPO_NAME
        Value:     examples
        Name:      REPO_OWNER
        Value:     kubeflow
      Image:       gcr.io/mlkube-testing/test-worker
      Volume Mounts:
        Mount Path:  /mnt/test-data-volume
        Name:        kubeflow-test-volume
        Mount Path:  /secret/github-token
        Name:        github-token
        Mount Path:  /secret/gcp-credentials
        Name:        gcp-credentials
      Working Dir:   /mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/examples
    Name:            py-lint
    Container:
      Command:
        python
        -m
        kubeflow.testing.prow_artifacts
        --artifacts_dir=/mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/output
        create_pr_symlink
        --bucket=kubernetes-jenkins
      Env:
        Name:   PYTHONPATH
        Value:  /mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/examples:/mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/testing/py
        Name:   GOPATH
        Value:  /mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/go
        Name:   GOOGLE_APPLICATION_CREDENTIALS
        Value:  /secret/gcp-credentials/key.json
        Name:   GIT_TOKEN
        Value From:
          Secret Key Ref:
            Key:   github_token
            Name:  github-token
        Name:      EXTRA_REPOS
        Value:     kubeflow/testing@HEAD
        Name:      BUILD_ID
        Value:     21
        Name:      BUILD_NUMBER
        Value:     21
        Name:      JOB_NAME
        Value:     kubeflow-examples-presubmit
        Name:      JOB_TYPE
        Value:     presubmit
        Name:      PULL_BASE_SHA
        Value:     f4fad9a8e0f8d27dc2fa07b4a6cd792ff26634ed
        Name:      PULL_NUMBER
        Value:     18
        Name:      PULL_PULL_SHA
        Value:     68c34529a3b785d7f22804776c13f29160efdeaa
        Name:      REPO_NAME
        Value:     examples
        Name:      REPO_OWNER
        Value:     kubeflow
      Image:       gcr.io/mlkube-testing/test-worker
      Volume Mounts:
        Mount Path:  /mnt/test-data-volume
        Name:        kubeflow-test-volume
        Mount Path:  /secret/github-token
        Name:        github-token
        Mount Path:  /secret/gcp-credentials
        Name:        gcp-credentials
      Working Dir:   /mnt/test-data-volume/kubeflow-examples-presubmit-examples-e2e-18-68c3452-21-047a/src/kubeflow/examples
    Name:            create-pr-symlink
  Volumes:
    Name:  github-token
    Secret:
      Secret Name:  github-token
    Name:           gcp-credentials
    Secret:
      Secret Name:  kubeflow-testing-credentials
    Name:           kubeflow-test-volume
    Persistent Volume Claim:
      Claim Name:  nfs-external
Status:
  Finished At:  2018-02-28T00:13:42Z
  Message:      invalid spec: spec.onExit template 'exit-handler' undefined
  Phase:        Failed
  Started At:   2018-02-28T00:13:42Z
Events:         <none>
```
Specifically the error is you are missing the template for 'exit-handler'

**Comment by texasmichelle** [comment]
Thanks for the help @jlewi! It appears to be working - is this a good starting place?

**Pull request merged**